<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Гасым! Давай знакомиться! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы». 


Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
    
Все ключевые этапы в работе выполнены, и я вижу что с проектом ты справшяешься. Есть моменты, которые нужно доработать, но я уверен, у тебя все получится.
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂
 
<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
</div>

<div class="alert alert-info">
<b> Гасым </b>


Привет, Дмитрий! Рад знакомству! Спасибо большое за твои детальные комментарии к каждой секции проекта, они очень ценны и помогли разобраться в непонятных для меня моментах и улучшить мою работу! Благодарю.

</div>

In [ ]:
!pip install -q transformers

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Лично меня очень утомляют простыни логов после установки библиотек. Я их все равно никто не читает. Можно добавить флаг -q  будет гораздо аккуратнее.
    
    !pip install -q transformers 
</div>

<div class="alert alert-info">
<b> Гасым </b>


Верно подмечено, меня также это утомляло, но не знал что так можно - теперь знаю, спасибо! 

</div>

In [ ]:
import pandas as pd
import numpy as np
import random
import matplotlib.pyplot as plt
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision.models as models
import torchvision.transforms as transforms
from torch import nn
from torchvision.models import ResNet50_Weights
from PIL import Image
from torchvision.models import resnet50
from transformers import BertTokenizer, BertModel
import os
import torch.optim as optim
from google.colab import drive
from sklearn.model_selection import GroupShuffleSplit
from sklearn.linear_model import LinearRegression
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
# Определим константу для пути к данным. Замените фактическим путем к вашей директории данных.
DATA_PATH = '/content/drive/My Drive'

# Подключимся к Google Drive - можно закомментировать, если запускаете локально!
drive.mount('/content/drive')

Mounted at /content/drive


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Очень "больная тема" - это загрузка данных и использование путей.Очевидно, что на моей машине (и скорее всего машине любого другого пользователя) пути к данным будут другие. И это "рабочая ситуация для большинства проектов. Решается она следующим образом:
 
    
  - Путь к директории с данными  должен быть задан в константе, в самом начале работы. Тогда ревьюеру (или любому другому пользователю твоего ноутбука, даже тебе самому, если нужно будет провести презентацию руководителю на другом компьютере) нужно будет изменить только этот путь.
  - Все остальные пути внутри ноутбука должны быть созданы с исопользованием этой константы, а не заданы жестко (есть сленговое определение "хардкод").  Более  того, настоятельно рекомендую для создания сложных путей вместо простой конкатерации строк с исползованием '/' использовать os.path.join(), тогда код будет одинаково надежно работать и в среде Windows и Linux (MacOS).
   
Чтобы запустить твой код при проверке я был вынужден руками менять код по всему ноутбуку. Прошу тебя внести изменения так, чтобы путь к данным нужно было изменять всего один раз.


А вот это образец универсального кода для загрузки данных    
    
    DATA_PATH = твой путь    
    
    df_train = pd.read_csv(os.path.join(DATA_PATH, 'train_dataset.csv'))
    ....
    
    
</div>

In [ ]:
# Создадим полные пути для экспертных и массовых аннотаций, используя DATA_PATH.
expert_annotations_path = os.path.join(DATA_PATH, 'ExpertAnnotations.tsv')
crowd_annotations_path = os.path.join(DATA_PATH, 'CrowdAnnotations.tsv')

# Загрузим датафреймы, используя полные пути
expert_annotations_df = pd.read_csv(expert_annotations_path, sep='\t', header=None)
crowd_annotations_df = pd.read_csv(crowd_annotations_path, sep='\t', header=None)

<div class="alert alert-info">
<b> Гасым </b>


Благодарю за полезный совет! Сделано. 

</div>

***Переименуем безымянные названия столбцов чтобы в дальнейшем было удобно по ним обьединить нужные столбцы в единый обучающий датасет.***

In [ ]:
# Именование столбцов файла CrowdAnnotations.tsv согласно приведенному описанию.
crowd_annotations_df.columns = [
    'image',
    'description_id',
    'percentage_of_matches',
    'number_of_confirmations',
    'number_of_rejections'
]

# Именование столбцов файла ExpertAnnotations.tsv согласно приведенному описанию.
expert_annotations_df.columns = [
    'image',
    'description_id',
    'expert_1_rating',
    'expert_2_rating',
    'expert_3_rating'
]

display(expert_annotations_df.head(10))
print()
display(crowd_annotations_df.head(10))

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
В наименованиях столбцов лучше убрать пробелы (можно заменить их на подчерки, а также использовать названия соответствующие snake_case)</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Я понимаю, что привычку "печатать" датафреймы с помощью print() прививает тренажер, но пожалуйста, не тяни это в ноутбук. Здесь лучше использовать display, вывод с его помощью выглядит гораздо аккуратнее.
    
Насколько я помню эту ошибку высвечивают еще в самом первом проекте, поэтому крайне обидно видеть её перед финальным проектом, только поэтому я подсвечиваю её красным. И конечно же прошу тебя исправить её во всем проекте,а не только в одной ячейке.    
</div>


<div class="alert alert-info">
<b> Гасым </b>


Дмитрий, благодарю за еще один полезный совет! К сожалению ни разу, ранее в проектах, мне не указывали на эту особенность и я впервые узнал о красивом выводе с помощью метода display(). Спасибо! 

</div>

In [ ]:
expert_annotations_df.info()
print()
crowd_annotations_df.info()

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
А вот оборачивать результат работы метода  .info() в функцию print() или display() это лишнее. Метод  info() и сам напечатает тебе нужную информацию. А вот через print()/display() ты выводишь только то,что данный метод возвращает - т.е. None
</div>

<div class="alert alert-info">
<b> Гасым </b>


Сделано! 

</div>

In [ ]:
# Нам необходимо соответствующим образом подготовить данные.
# Мы преобразуем рейтинги экспертов в шкалу от 0 до 1, чтобы соответствовать шкале процента краудсорса.

# Нормализуем экспертные оценки по шкале 0-1
expert_annotations_df['Normalized Expert 1'] = (expert_annotations_df['expert_1_rating'] - 1) / 3
expert_annotations_df['Normalized Expert 2'] = (expert_annotations_df['expert_2_rating'] - 1) / 3
expert_annotations_df['Normalized Expert 3'] = (expert_annotations_df['expert_3_rating'] - 1) / 3

# Рассчитаем средний экспертный рейтинг
expert_annotations_df['mean_expert_rating'] = expert_annotations_df[
    ['Normalized Expert 1', 'Normalized Expert 2', 'Normalized Expert 3']].mean(axis=1)

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Чтобы привести к шкале 0-1 я бы посоветовал использовать линейное преобразование "минус 1 разделить на 3". Если использовать деление на 4, то минимальное значение будет 0,25
</div>

<div class="alert alert-info">
<b> Гасым </b>


Сделано! 

</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
   
    
В этом проекте есть несколько возможных стратегий работы с обучающими и краудсорс оценками,  вот эти стратегии с плюсами и минусами:
   
    
 - использовать соединение через outer, в этом случае, при корректной обработке пропусков мы можем получить более 50К размеченых пар. К плюсам данного подхода можно отнести большое количество данных. Минус - в основном это будут данные, размеченные краудсорсерами, а там качество раметки ниже.
 - использовать только экспертные оценки. плюс - высокое качество данных (разметка имеет шкалу). Минус - данных меньше.  Поясню по поводу качества разметки на примере: если на изображении одна собака, а в описании "две собаки бегут по берегу", то эксперты поставять соответствие 0,6-0,7 (переводя в шкалу 0-1), а краудсорсеры поставят 0.  Но на экспертных оценках модель хотя-бы научится находить собак, а на краудсор оценках ничему не начится.
    
 - использование соединения через left.  Данный подход очевидно проигрывает второму варианту: данных столько же, нужно дополнительно возиться с объединением, но качество разметки снижается.. Но это соответсвует предложениям авторов проекта, поэтому такой подход имеет право на жизнь
    
 - объединение через inner. Тут все плохо - и данных ОЧЕНЬ мало (в два раза меньше, чем использовать только экспертные оценки) и качество у них "подпорчено". И навыков по обработке пропусков не получаем.....
    
Сейчас ты соединил экспертные и краудсорс оценки, через1 `inner`, и мы просто потеряли половину данных, не получив ничего в замен (точнее только ухудшив наши оценки).
    
Пожалуйста используй любую другую стратегию, кроме внутреннего соединения. Слишком мало данных при этом получается.    
    


</div>

<div class="alert alert-info">
<b> Гасым </b>


Спасибо что указал на этот критически важный момент. Выбрал первую стратегию с outer соединением. Сделано! 

</div>

In [ ]:
# Присвоим веса экспертным и краудсорсинговым аннотациям.
expert_weight = 0.6
crowd_weight = 0.4

# Выполнить внешнее соединение (outer merge)
merged_annotations_outer = pd.merge(
    crowd_annotations_df,
    expert_annotations_df,
    how='outer',
    on=['image', 'description_id']
)

# Расчет медианы краудсорса
median_crowd_percentage = merged_annotations_outer['percentage_of_matches'].median()
# Расчет медианы экспертов
median_expert_rating = merged_annotations_outer['mean_expert_rating'].median()

# Заполним пропущенные значения медианой
merged_annotations_outer['normalized_crowd_percentage'] = merged_annotations_outer[
    'percentage_of_matches'].fillna(median_crowd_percentage)
merged_annotations_outer['mean_expert_rating'] = merged_annotations_outer[
    'mean_expert_rating'].fillna(median_expert_rating)

# Рассчитаем взвешенный комбинированный балл
merged_annotations_outer['weighted_combined_score'] = (
    (merged_annotations_outer['mean_expert_rating'] * expert_weight) +
    (merged_annotations_outer['normalized_crowd_percentage'] * crowd_weight)
)

# Выберем только необходимые столбцы и назначим их датафрейму annotations_df.
annotations_df = merged_annotations_outer[['image', 'description_id', 'weighted_combined_score']]

# Отобразим первые несколько строк annotations_df для проверки.
annotations_df.head()

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Старайся не делать такие длинные строки. Стандарт PEP8 регулирует длину строки 79 символов. Придерживаться такого очень сложно, но лучше хотя-бы вмещать в экран, чтобы не приходилось использовать scroll.
    
    
А еще  если сомневаешься, как отформатировать правильно - всегда можно найти какой-нибудь оналйн-форматтер, например этот
    
  https://extendsclass.com/python-formatter.html
    
    
В качестве дополнительного развития могу порекомендоваь книгу "PYTHON  чистый код для продолжающих", Эл Свейгарт    
    
   
    
    
</div>

<div class="alert alert-info">
<b> Гасым </b>


Принято! 

</div>

In [ ]:
annotations_df.info()

In [ ]:
# Чтобы понять сокращение количества записей, мы проверим охват и согласованность 
# 'image' и 'description_id' как в датафрейме краудсорса, так и в экспертных аннотациях.

# Количество уникальных 'image' и 'description_id'  в обоих наборах данных
unique_crowd_images = crowd_annotations_df['image'].nunique()
unique_expert_images = expert_annotations_df['image'].nunique()
unique_crowd_descriptions = crowd_annotations_df['description_id'].nunique()
unique_expert_descriptions = expert_annotations_df['description_id'].nunique()

# Проверка любых расхождений в форматировании или проблем с сопоставлением
# Мы посмотрим на пересечения наборов уникальных идентификаторов, чтобы увидеть, 
# сколько из них являются общими.
shared_image_files = set(
    crowd_annotations_df['image']).intersection(set(expert_annotations_df['image']))
shared_description_ids = set(
    crowd_annotations_df['description_id']).intersection(set(expert_annotations_df['description_id']))

# Подсчитаем количество общих идентификаторов
num_shared_image_files = len(shared_image_files)
num_shared_description_ids = len(shared_description_ids)

print(unique_crowd_images, unique_expert_images, unique_crowd_descriptions, 
      unique_expert_descriptions, num_shared_image_files, num_shared_description_ids)

(1000, 1000, 1000, 977, 1000, 977)

In [ ]:
# 1. Распределение взвешенного комбинированного балла
score_description = annotations_df['weighted_combined_score'].describe()
score_histogram = annotations_df['weighted_combined_score'].hist()
plt.title('Distribution of Weighted Combined Scores')
plt.xlabel('Weighted Combined Score')
plt.ylabel('Frequency')
plt.grid(False)
plt.show()

# 2. Разнообразие изображений
unique_images = annotations_df['image'].nunique()

# 3. Разнообразие описаний
unique_descriptions = annotations_df['description_id'].nunique()

score_description, unique_images, unique_descriptions

На представленном графике показаны результаты процесса анализа данных, включая распределение «взвешенных комбинированных оценок» и количество уникальных имен файлов изображений и идентификаторов описаний по двум наборам данных (аннотации, полученные из краудсорсинга, и экспертные аннотации). Вот интерпретация отображаемых результатов:

### Распределение взвешенных комбинированных баллов
Гистограмма иллюстрирует распределение «взвешенных комбинированных оценок», которые являются показателем, полученным как на основе краудсорсинговых, так и экспертных оценок. Вывод метода `.describe()` предоставляет сводную статистику этих оценок:
- **count**: В наборе данных 51.323 записи, что указывает на значительный объем данных для анализа и обучения модели.
- **Среднее**: средний балл составляет примерно 0.099.
- **Стандартное отклонение (стандартное)**: стандартное отклонение оценок составляет примерно 0.112, что указывает на умеренную вариабельность оценок.
- **Минимум (мин)**: Минимальный балл равен 0, что указывает на то, что существуют пары изображений и описаний, которые полностью не соответствуют согласно оценкам.
- **25-й процентиль**: 25 % оценок ниже примерно 0.067.
- **50-й процентиль (медиана)**: средний балл составляет 0.067.
- **75-й процентиль**: 75 % оценок ниже примерно 0.067. 
- **25-й, 50-й (медиана) и 75-й процентили равны примерно 0,067**. Такая однородность по квартилям указывает на концентрацию баллов вокруг этого значения. Согласованность этих значений процентилей предполагает, что значительная часть набора данных имеет одинаковый уровень соответствия между изображениями и описаниями в соответствии с системой взвешенной оценки.
- **Максимум (макс.)**: Максимальный балл равен 1, что показывает, что некоторые пары «изображение-описание» идеально совпадают в соответствии с совокупными оценками толпы и экспертов.

**Анализ гистограммы и последствия для моделирования**

- ***Чувствительность модели***: модель должна быть достаточно чувствительной, чтобы различать различные степени соответствия, тем более что большая часть данных сосредоточена вокруг медианного значения.
- ***Обработка асимметрии***: Правильная асимметрия данных может повлиять на способность модели точно прогнозировать более высокие оценки. При рассмотрении таких случаев может потребоваться особое внимание.
- ***Представление данных***: гистограмма показывает, что данные предоставляют хороший диапазон сценариев: от отсутствия совпадения до идеального совпадения, что полезно для обучения надежной модели.

В целом гистограмма указывает на разумное распределение оценок со значительной концентрацией вокруг медианы, подходящее для обучения и оценки модели, целью которой является сопоставление изображений с описаниями.

### Анализ идентификатора изображения и описания
- **Уникальные изображения толпы**: 1000 уникальных имен файлов изображений в наборе данных, полученном из краудсорсинга.
- **Уникальные экспертные изображения**: 1000 уникальных имен файлов изображений в наборе экспертных данных.
- **Уникальные описания толпы**: 1000 уникальных идентификаторов описаний в наборе краудсорсинговых данных.
- **Уникальные описания экспертов**: 977 уникальных идентификаторов описаний в наборе экспертных данных.
- **Общие файлы изображений**: все 1000 имен файлов изображений используются совместно краудсорсинговыми и экспертными наборами данных.
- **Общие идентификаторы описаний**: в двух наборах данных имеется 977 общих идентификаторов описаний.

### Интерпретация:
- **Согласованность между наборами данных**. В именах файлов изображений между краудсорсинговыми и экспертными наборами данных наблюдается высокий уровень согласованности, при этом все файлы изображений присутствуют в обоих. Однако существует небольшое расхождение в идентификаторах описаний: в наборе краудсорсинговых данных присутствует 23 уникальных описания, которых нет в наборе экспертных данных.
- **Описательная статистика баллов**: взвешенные комбинированные баллы смещены в сторону нижнего предела шкалы, при этом медиана ниже среднего значения. Это указывает на то, что больше аннотаций имеют оценку ниже среднего.
- **Последствия для дальнейшего анализа**: наличие двух пиков в распределении баллов требует дальнейшего изучения. Возможно, было бы полезно понять характеристики аннотаций, получивших высокие и низкие оценки, и коррелируют ли они с конкретными особенностями изображений или описаний.
- **Полнота данных**. Расхождение в идентификаторах описаний позволяет предположить, что краудсорсинговый набор данных может содержать дополнительные аннотации, не оцененные экспертами. Дальнейшее изучение этих дополнительных описаний могло бы дать представление о том, представляют ли они пробел в экспертном анализе или были исключены по определенным причинам.

Этот анализ необходим для понимания согласия между различными источниками аннотаций и качества аннотаций. Он может направлять разработку моделей, основанных на этих аннотациях, принимать решения о том, каким аннотациям доверять или отдавать приоритет, а также предлагать области, в которых сам процесс аннотирования может быть улучшен или дополнительно изучен.

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец, что не забыаешь про выводы (подведение итогов) после исследования.    
    
-----    
Кстати, есть еще достаочно интересный источник данных. Зная структуру query_id (точнее то, что там содержится ссылка на изображение для которого оно было написано), можно получить 977 пар изображение/описание идеально подходящих друг к другу.</div>

# Проверка данных

In [ ]:
# Загрузим датасет
file_path = os.path.join(DATA_PATH, 'train_dataset.csv')
train_dataset = pd.read_csv(file_path)

# Отобразим первые несколько строк набора данных и его основную информацию.
train_dataset.info()
print()
display(train_dataset.head())

In [ ]:
# Определим список запрещенных ключевых слов.
restricted_keywords = ['minor', "child", "children", "toddlers",
                       "kid", "boy", "girl", "school", "playground"]

# Функция проверки наличия в описании каких-либо запрещенных ключевых слов.
def contains_restricted_keyword(text, keywords):
    text = text.lower()
    return any(keyword in text for keyword in keywords)

# Шаг 1. Выявим плохие комментарии
train_dataset['is_bad_comment'] = train_dataset['query_text'].apply(
    lambda x: contains_restricted_keyword(x, restricted_keywords))

# Шаг 2. Определим список плохих изображений на основе плохих комментариев.
# Отрезаем два последних символа query_id плохих комментариев, чтобы получить список плохих изображений
bad_images = set(train_dataset[train_dataset['is_bad_comment']]['query_id'].apply(lambda x: x[:-2]))

# Шаг 3. Удаляем из нашего набора данных все пары изображений и описаний с плохими изображениями.
filtered_train_df = train_dataset[~train_dataset['image'].isin(bad_images)]

# Отображение результата фильтрации
original_count = len(train_dataset)
filtered_count = len(filtered_train_df)

print(original_count, filtered_count)
print()
filtered_train_df.info()
print()
display(filtered_train_df.head())

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
к сожалению читать результат такого вывода невозможно.. пожалуйста используй print для чисел и display для датафреймов.
    
    
</div>


<div class="alert alert-info">
<b> Гасым </b>


Исправил! 

</div>

In [ ]:
# Шаг 1. Переименуем столбцы в filtered_train_df, чтобы они соответствовали annotations_df.
filtered_train_df.rename(columns={'query_id': 'description_id'}, inplace=True)

# Шаг 2. Объединим annotations_df с filtered_train_df
pre_train_df = pd.merge(annotations_df, filtered_train_df, on=['image', 'description_id'])

# Проверка результата
pre_train_df.info()
print()
display(pre_train_df.head())

In [ ]:
pre_train_df = pre_train_df[['image', 'description_id',
                             'query_text', 'weighted_combined_score']]
pre_train_df.info()
print()
display(pre_train_df.head())

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
    
Ты удалил комментарии, в которых были запрещенные слова. Но наша задача убрать изображения, содержащие детей. Есть два способа сделать это:
    
 - Определить список плохих изображений, как изображения удовлетворяющие условиям:
    - Комментарий содержит плохие слова
    - <b>Оценка соответствия комментария и изображения выше порога.</b>
 - Удалить из нашего датасета все пары "изображение/описание" с плохими изображениями.
    
 
И второй, очень красивый способ. Он основан на том, что `query_id` содержит в с себе имя изображения, для которого он был написан (такая пасхалочка от авторов датасета).
    
 - Определить список плохих комментариев
 - У `query_id`  плохих комментариев отрезать два последних символа и получим список плохих изображений.
    
    
    
    
    
</div>

<div class="alert alert-info">
<b> Гасым </b>

Спасибо большое за подсказку, Дмитрий! Сделано. 

</div>

# Векторизация текстов

In [ ]:
# Функция кодирования текста с использованием BERT с помощью графического процессора
def bert_encode(text, model, tokenizer, device):
    input_ids = tokenizer.encode(text, add_special_tokens=True)
    input_ids = torch.tensor(input_ids).unsqueeze(0).to(device)

    # Получим выходные данные (последние скрытые состояния) из модели BERT.
    with torch.no_grad():
        last_hidden_states = model(input_ids)[0]

    # Переместим результат обратно в ЦП и преобразуем в numpy
    return last_hidden_states[0][0].cpu().numpy()

# Проверим доступность графического процессора
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Загрузим предварительно обученный токенизатор модели и модель и отправим модель в графический процессор.
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased').to(device)

# Векторизуем текстовые описания
pre_train_df['bert_vector'] = pre_train_df['query_text'].apply(lambda x: bert_encode(x, model, tokenizer, device))

Using device: cuda


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:72: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

In [ ]:
pre_train_df.head()

Image file name               Description ID  \
0  1056338697_4f7d7ce270.jpg   434792818_56375e203f.jpg#2   
1  1082379191_ec1e53f996.jpg  1536774449_e16b1b6382.jpg#2   
2  1084040636_97d9633581.jpg   256085101_2c2617c5d0.jpg#2   
3  1084040636_97d9633581.jpg  3396157719_6807d52a81.jpg#2   
4  1096395242_fc69f0ae5a.jpg  1425069308_488e5fcf9d.jpg#2   

                                          query_text  Weighted Combined Score  \
0  A man and woman look back at the camera while ...                 0.200000   
1              A crowd of people at an outdoor event                 0.300000   
2       Close-up of dog in profile with mouth open .                 0.533333   
3                       A dog gets on a trampoline .                 0.250000   
4  A white dog jumps to catch a blue ball in a li...                 0.300000   

                                         bert_vector  
0  [0.20837602, 0.19895823, 0.058023393, -0.02440...  
1  [-0.014534058, -0.039590366, -0.2054309, 0.139...  
2  [-0.78212327, -0.5189205, -0.44774392, 0.03597...  
3  [-0.17570308, 0.104248144, -0.6895283, 0.16900...  
4  [-0.36722273, 0.01828046, -0.89691925, 0.20806...

In [ ]:
pre_train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1811 entries, 0 to 1810
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Image file name          1811 non-null   object 
 1   Description ID           1811 non-null   object 
 2   query_text               1811 non-null   object 
 3   Weighted Combined Score  1811 non-null   float64
 4   bert_vector              1811 non-null   object 
dtypes: float64(1), object(4)
memory usage: 84.9+ KB


In [ ]:
# Выберем один вектор из датафрейма, чтобы проверить его размерность.
sample_vector = pre_train_df['bert_vector'].iloc[0]

# Распечатаем размерность вектора выборки
print(f"Размерность вектора BERT равна: {sample_vector.shape}")

Чтобы векторизировать тексты с помощью BERT из Hugging Face Transformers, были выполнены следующие действия:

1. **Настройка среды**: установлена необходимая библиотека из Hugging Face - transformers.
2. **Загрузка предварительно обученной модели BERT и токенизатора**: Использована предварительно обученная модель BERT, подходящая для наших нужд, вместе с соответствующим токенизатором.
3. **Предварительная обработка и токенизация текста**: Подготовлены текстовые данные для модели, включая их токенизацию и правильное форматирование.
4. **Генерация эмбеддингов**: токенизированный текст был передан модели, чтобы получить эмбеддинги - числовые векторы.

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично,полученные векторные представления для текстов описаний с помощью BERT </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Но было бы неплохо показать размерность полученного вектора описания
</div>

<div class="alert alert-info">
<b> Гасым </b>

Отличная идея! Сделано. 

</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
В этом раздеел есть проблема и заключается она в следующем. Для векторизации ты использовал BERT, основанный на Pytorch.  А для векторизации изображений и итоговой модели используешь Keras (pytorch).  Поддержка двух DL фреймворков - это будет ад для девопса. Поэтому хорошей практикой является использование одного фреймворка в одном проекте. Можно взять TF версию модели из tramnsformers (или модель BERT из Keras NLP)

----
    
Хотя должен отметить, что после выхода последней версии Keras, поддерживающей pytorch границы между фрейморками начинают стираться.    


</div>

<div class="alert alert-info">
<b> Гасым </b>

Сделано! В векторизации изображений применил PyTorch (как и при векторизации текстов) вместо TF Keras.

</div>

# Векторизация изображений

In [ ]:
image_dir = os.path.join(DATA_PATH, 'train_images')

In [ ]:
# Определите устройство
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Загрузим предварительно обученную модель ResNet50.
resnet_model = models.resnet50(pretrained=True)
# Удалим последний полностью связанный слой
modules = list(resnet_model.children())[:-1]
resnet_model = nn.Sequential(*modules)
# Переместим модель на устройство и установим ее в оценочный режим.
resnet_model = resnet_model.to(device)
resnet_model.eval()

# Определите преобразования
transform = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])

# Функция кодирования изображений с помощью ResNet50
def resnet50_encode(img_path):
    img = Image.open(img_path)
    img_t = transform(img).unsqueeze(0).to(device)

    with torch.no_grad():
        features = resnet_model(img_t)

    return features.cpu().numpy().flatten()

# Инициализируем список для хранения имен файлов изображений и соответствующих им векторов.
image_data = []
# Обработка каждого изображения в директории
for img_name in os.listdir(image_dir):
    img_path = os.path.join(image_dir, img_name)
    if os.path.isfile(img_path):  # Убедимся, что это файл, а не подкаталог.
        img_vector = resnet50_encode(img_path)
        image_data.append({'image': img_name, 'resnet50_vector': img_vector})

# Преобразуем список словарей в DataFrame
image_vectors_df = pd.DataFrame(image_data)
image_vectors_df

In [ ]:
image_vectors_df.info()

In [ ]:
# Объединение pre_train_df с image_vectors_df по 'Image file name'
final_train_df = pd.merge(pre_train_df, image_vectors_df, on='image')
final_train_df.info()

In [ ]:
final_train_df.head()

In [ ]:
# Выберем один вектор из датафрейма, чтобы проверить его размерность.
sample_image_vector = final_train_df['resnet50_vector'].iloc[0]

# Распечатаем размер образца векторного изображения.
print(f"Размерность вектора изображения ResNet50 равна: {sample_image_vector.shape}")

Учитывая размеры векторов изображения и текста, полученных из ResNet50 и BERT соответственно, мы можем детализировать полученные пары векторов следующим образом:

1. **Размер вектора изображения ResNet50**: (2048,)
    — Каждое изображение представлено вектором длиной 2048, который является результатом модели ResNet50 после обработки изображения.

2. **Размер вектора текста BERT**: (768,)
    — Каждое текстовое описание представлено вектором длиной 768, полученным из модели BERT.

### Объединенный вектор для входных данных модели

Для создания модели, которая оценивает близость изображения и текстового описания, нам необходимо объединить эти два вектора. Полученный объединенный вектор будет иметь следующие свойства:

- **Размер конкатенированного вектора**: (2048 + 768) = (2816)
    - Это получается простым сложением размеров отдельных векторов.
    - Объединенный вектор эффективно сочетает в себе признаки, извлеченные как из изображения, так и из текстового описания.

### Использование в модели

- Объединенный вектор размерности (2816) будет использоваться в качестве входных данных для модели машинного обучения.
- Эта модель научится прогнозировать итоговую оценку экспертов на основе совокупности признаков векторов текста и изображения.
- Модель может представлять собой линейную регрессию, полностью связанную нейронную сеть или любую другую подходящую модель, способную обрабатывать эту размерность и выполнять задачи регрессии.

### Заключение

Комбинированный вектор изображений и текстовых представлений обеспечивает полный набор признаков, на которых модель может учиться. Он инкапсулирует как визуальную, так и текстовую информацию, что имеет решающее значение для точного прогнозирования экспертной оценки соответствия изображения и его описания. Размерность (2816) должна быть приемлемой для большинства современных сред машинного обучения и обеспечивать эффективное обучение и прогнозирование.

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Здесь все ОК, векторизация изображений произведена верно </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Но было бы неплохо показать размерность полученного вектора изображения
</div>

<div class="alert alert-info">
<b> Гасым </b>

Сделано. 

</div>

# Обучение модели

In [ ]:
# Конкатенируем «bert_vector» и «resnet50_vector», чтобы создать один вектор признаков.
X = np.hstack((np.array(final_train_df['bert_vector'].tolist()),
               np.array(final_train_df['resnet50_vector'].tolist())))

# Целевая переменная 'Weighted Combined Score'
y = final_train_df['weighted_combined_score'].values

# Определим группы
groups = final_train_df['image'].values

# Инициализируем GroupShuffleSplit
gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)

# Получим индексы для первого разделения
train_idx, valid_idx = next(gss.split(X, y, groups=groups))

# Разделение данных на обучающие и тестовые наборы с использованием индексов
X_train, X_valid = X[train_idx], X[valid_idx]
y_train, y_valid = y[train_idx], y[valid_idx]

print(X_train.shape, X_valid.shape, y_train.shape, y_valid.shape)

**Вывод**:

- Объединенный вектор признаков X имеет размерность (2816) и объединяет 768-мерный текстовый вектор BERT и 2048-мерный вектор изображения ResNet50.
- GroupShuffleSplit используется для обеспечения отсутствия одного и того же изображения как в обучающем, так и в проверочном наборах, обеспечивая целостность процесса проверки.
- Размеры наборов обучения и проверки подтверждают, что данные были правильно разделены.  Обучающий набор содержит 3311 выборок, а проверочный набор — 834 выборки, каждая из которых имеет векторы признаков 2816 измерений.

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 Вектора, полученные из текста и изображения объеденены👍 </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Здесь напрашивается небольшой вывод, хотя бы с указанием размерностей полученных признаков.  
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, благодаря использованию GroupShuffleSplit ты исключаешь попадание одного изображения в обучающий и валидационный наборы.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Только давай разберемся с выборками. Финальное тестирование (визуальное) мы делаем на отложенной выборке. Поэтому выборка,которую ты сейчас отделил играет роль валидационной. Есть смысл в её названии использовать слово valid а не test</div>

<div class="alert alert-info">
<b> Гасым </b>

Сделано. 

</div>

## Модель 1: Baseline DummyRegressor

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
В этом задании требуется выбрать основную метрику, которую ты будешь оптимизировать. Я вижу что ты уже выбрал метрики  mse/rmse, поэтому просто напиши обоснование для своего выбора.  </div>

<div class="alert alert-info">
<b> Гасым </b>

Сделано. 

</div>

**Обоснование выбора MSE/RMSE**:

1. ***Характер проблемы***: Поскольку мы прогнозируем непрерывную переменную (взвешенный комбинированный балл), подходят такие метрики регрессии, как MSE и RMSE.  


2. ***Интерпретируемость***:
    - **MSE**: измеряет среднеквадратическую разницу между расчетными значениями и фактическими значениями. Будучи возведенным в квадрат, он наказывает за большие ошибки более строго, чем за меньшие, что часто желательно в задачах регрессии.
    - **RMSE**: это просто квадратный корень из MSE, который возвращает метрику ошибки к тому же масштабу, что и целевая переменная. Это делает RMSE более интерпретируемым, поскольку его можно понимать в тех же единицах, что и прогнозируемую переменную.  


3. ***Чувствительность к выбросам***: И MSE, и RMSE чувствительны к выбросам, что может быть полезно, если в нашем контексте важно выявлять и наказывать большие отклонения от фактических значений.

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Мы решаем новую задачу и не знаем, наксколько хороши наши метрики и наши модели. Поэтому советую начать с бейзлайна - константной модели (можешь взять простой DummyRegressor).</div>

Прежде чем углубляться в более сложные модели, рекомендуется определить базовую производительность с помощью простой модели. Это помогает понять ожидаемую минимальную производительность и оценить улучшение, предлагаемое более сложными моделями. DummyRegressor из scikit-learn — хороший выбор для этой цели:

In [ ]:
# Инициализируем DummyRegressor
dummy_regr = DummyRegressor(strategy="mean")

# Подгоним DummyRegressor к обучающим данным.
dummy_regr.fit(X_train, y_train)

# Прогнозирование на основе данных проверки
y_dummy_pred = dummy_regr.predict(X_valid)

# Оценка
mse_dummy = mean_squared_error(y_valid, y_dummy_pred)
rmse_dummy = np.sqrt(mse_dummy)

print("Dummy Regressor - MSE:", mse_dummy)
print("Dummy Regressor - RMSE:", rmse_dummy)

## Модель 2: Линейная регрессия

In [ ]:
# Инициализируем и обучим модель линейной регрессии
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

# Прогнозирование на основе тестовых данных
y_pred_lr = lr_model.predict(X_valid)

# Оценка модели
mse_lr = mean_squared_error(y_valid, y_pred_lr)
rmse_lr = np.sqrt(mse_lr)

print("Linear Regression - MSE:", mse_lr)
print("Linear Regression - RMSE:", rmse_lr)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Есть линейная модель. И видно что она чудовищно переобучается. </div>

## Модель 3: Полносвязная нейронная сеть (с использованием PyTorch)

In [ ]:
# Определим модель нейронной сети с помощью PyTorch
class NeuralNetwork(nn.Module):
    def __init__(self, input_size):
        super(NeuralNetwork, self).__init__()
        self.fc1 = nn.Linear(input_size, 256)
        self.fc2 = nn.Linear(256, 128)
        self.fc3 = nn.Linear(128, 64)
        self.fc4 = nn.Linear(64, 1)

    def forward(self, x):
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = F.relu(self.fc3(x))
        x = self.fc4(x)  # Linear output for regression
        return x

# Инициализируем модель
input_size = X_train.shape[1]
nn_model = NeuralNetwork(input_size)
nn_model = nn_model.to(device)

In [ ]:
# Определим оптимизатор и функцию потерь
optimizer = optim.Adam(nn_model.parameters(), lr=0.0001)
loss_fn = nn.MSELoss()

# Преобразование массивов numpy в тензоры torch
X_train_torch = torch.tensor(X_train).float().to(device)
y_train_torch = torch.tensor(y_train).float().to(device)
X_valid_torch = torch.tensor(X_valid).float().to(device)
y_valid_torch = torch.tensor(y_valid).float().to(device)

num_epochs = 100

# Обучающий цикл
for epoch in range(num_epochs):
    nn_model.train()
    optimizer.zero_grad()
    outputs = nn_model(X_train_torch)
    loss = loss_fn(outputs.squeeze(), y_train_torch)
    loss.backward()
    optimizer.step()

    # Валидация
    nn_model.eval()
    with torch.no_grad():
        valid_outputs = nn_model(X_valid_torch)
        valid_loss = loss_fn(valid_outputs.squeeze(), y_valid_torch)

    # Печать для каждой 10-й эпохи
    if (epoch + 1) % 10 == 0:
        print(f"Epoch {epoch+1}, Loss: {loss.item()}, Validation Loss: {valid_loss.item()}")

In [ ]:
# Прогноз
nn_model.eval()
with torch.no_grad():
    y_pred_nn = nn_model(X_valid_torch).cpu().numpy().flatten()

# Оценка
mse_nn = mean_squared_error(y_valid, y_pred_nn)
rmse_nn = np.sqrt(mse_nn)

print("Neural Network - MSE:", mse_nn)
print("Neural Network - RMSE:", rmse_nn)

На основании предоставленных показателей производительности моделей «Dummy Regressor», «линейная регрессия» и «нейронная сеть» мы можем сделать несколько выводов:

1. **Производительность Dummy Regressor**:
    - **MSE**: 0,069477
    - **RMSE**: 0,263585
    - Эта модель служит базовой линией, просто предсказывая среднее значение зависимой переменной, наблюдаемой в обучающих данных. Здесь MSE и RMSE являются эталонами, по которым измеряются характеристики других моделей.

2. **Производительность линейной регрессии**:
    - **MSE**: 0,094267
    - **RMSE**: 0,307029
    - Линейная регрессия работает хуже, чем Dummy Regressor, о чем свидетельствуют более высокие значения MSE и RMSE. Это говорит о том, что простой линейный подход может оказаться недостаточным для выявления сложностей или закономерностей в данных.

3. **Производительность нейронной сети**:
    - **MSE**: 0,060443
    - **RMSE**: 0,245851
    - Нейронная сеть демонстрирует значительное улучшение по сравнению с моделями Dummy Regressor и линейной регрессии. Более низкие значения MSE и RMSE указывают на лучшую точность прогнозирования и эффективность модели в выявлении нелинейных связей в данных.

### Общие выводы:

- **Сложность модели**. Улучшенная производительность нейронной сети предполагает, что взаимосвязь между признаками и целевой переменной является нелинейной и сложной, что нейронная сеть способна лучше уловить по сравнению с линейной регрессией.

- **Проблемы переобучения**. Несмотря на более высокую производительность нейронной сети, всегда существует опасность переобучения более сложными моделями, особенно при работе с многомерными данными. Важно убедиться, что модель хорошо обобщает невидимые данные, возможно, используя такие методы, как перекрестная проверка, регуляризация или dropout (в случае нейронных сетей).

- **Дальнейшая оптимизация**: хотя нейронная сеть показала многообещающие результаты, дальнейшая настройка гиперпараметров (например, количества слоев, количества нейронов, скорости обучения и т. д.) или экспериментирование с различными архитектурами потенциально может дать еще лучшие результаты.

- **Качество данных и признаки**. Качество и актуальность признаков играют решающую роль в производительности модели. Возможно, будет полезно вернуться к проектированию и выбору признаков, чтобы посмотреть, можно ли еще улучшить производительность модели.

Таким образом, нейронная сеть кажется наиболее многообещающей моделью среди протестированных, предлагая лучшую производительность с точки зрения MSE и RMSE. Однако важно учитывать сложность модели и необходимость тщательной проверки для обеспечения надежных и обобщаемых результатов.

***Однако этого может быть достаточно на данном экспериментальном этапе в соответствии с задачей PoC (Proof of Concept, Проверка концепции)!***

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
В целом по нейронке все ОК, но оставлю советы, куда бы я копал для улучшения метрики
    
 - борьба с переобучением. Линейная регрессия показала, что на на наших данных модели могут переобучаться. Ты взял сложную модель, которая еще больше может быть склонной к переобучению. Чтобы модель научилась чему-то толковому может понадобиться больше чем 10 эпох, но тогда нужно позаоботиться о борьбе с переобучением.
    
 - ты использовал дефолтный adam с lr 1e-3, но иногда возникает необходимость уменьшить/увеличить шаг обучения
 - по моему опыту 10 эпох это слишком мало, успешные модели обучались сотни и даже тысячи эпох, прежде чем научиться чему-то толковому
 - используй доступную тебе валидационную выборку, не нужно дополнительно терять 10% данных на validation_split 
    
    .</div>

# Тестирование модели и демонстрация ее работы

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Здесь я попрошу тебя немного переделать логику инференса.
    
Ты уже на финишной прямой курса и во всю ипользуешь функции. Поэтому я попрошу тебя оформить тестирование следующим образом: нужно написать функцию, способную искать изображение по тексту для одного изображения.
 
  Что должна делать функция
    
   - Получать текст
   - Возвращать картинку и метрику близости. 
   - Выдавать дисклеймер если запрос некорректный
    
А затем нужно 10 раз применить функцию к 10 случайным запросам.    
    

------------
    
Исходя из описанной мной логики векторизацию изображений оптимально выполнить заранее, а вот запрос лучше векторизовать "на лету".    


</div>

In [ ]:
file_path = os.path.join(DATA_PATH, 'test_queries.csv')

# Попытка загрузить первые несколько строк для проверки данных
try:
    test_queries_df = pd.read_csv(file_path, delimiter='|', on_bad_lines='skip')
    error_message = "File read with specific delimiter and skipping bad lines."
except Exception as e:
    test_queries_df = None
    error_message = str(e)

test_queries_df.info()
print()
display(test_queries_df.head())

In [ ]:
test_queries_df = test_queries_df.drop('Unnamed: 0', axis=1)
test_queries_df.head()

query_id  \
0  1177994172_10d143cb8d.jpg#0   
1  1177994172_10d143cb8d.jpg#1   
2  1177994172_10d143cb8d.jpg#2   
3  1177994172_10d143cb8d.jpg#3   
4  1177994172_10d143cb8d.jpg#4   

                                          query_text  \
0  Two blonde boys , one in a camouflage shirt an...   
1  Two boys are squirting water guns at each other .   
2            Two boys spraying each other with water   
3  Two children wearing jeans squirt water at eac...   
4  Two young boys are squirting water at each oth...   

                       image  
0  1177994172_10d143cb8d.jpg  
1  1177994172_10d143cb8d.jpg  
2  1177994172_10d143cb8d.jpg  
3  1177994172_10d143cb8d.jpg  
4  1177994172_10d143cb8d.jpg

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Векторизацию описаний не стоит делать заранее
  
</div>

In [ ]:
test_queries_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   query_id     500 non-null    object
 1   query_text   500 non-null    object
 2   image        500 non-null    object
 3   bert_vector  500 non-null    object
dtypes: object(4)
memory usage: 15.8+ KB


In [ ]:
test_images_dir = os.path.join(DATA_PATH, 'test_images')

image_data = []
for img_name in os.listdir(test_images_dir):
    # Пропускаем любой файл, который не заканчивается на .jpg.
    if not img_name.lower().endswith('.jpg'):
        continue
    
    img_path = os.path.join(test_images_dir, img_name)
    if os.path.isfile(img_path):
        img_vector = resnet50_encode(img_path)
        image_data.append({'image': img_name, 'resnet50_vector': img_vector})

# Преобразование списка словарей в DataFrame
test_images_df = pd.DataFrame(image_data)
display(test_images_df.head())
print()
test_images_df.info()

In [ ]:
# Объединение двух датафреймов по столбцу «image»
test_df = test_queries_df.merge(test_images_df, on='image', how='inner')
test_df.info()
print()
display(test_df.head())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   query_id         500 non-null    object
 1   query_text       500 non-null    object
 2   image            500 non-null    object
 3   bert_vector      500 non-null    object
 4   resnet50_vector  500 non-null    object
dtypes: object(5)
memory usage: 23.4+ KB
None


In [ ]:
def find_similar_image(query, test_df, nn_model, bert_model, tokenizer, resnet_model, device):
    # Векторизация текста запроса с помощью BERT
    query_vector = bert_encode(query, bert_model, tokenizer, device)

    max_score = -1
    best_image = None

    # Итерация через каждый вектор изображения в наборе тестовых данных.
    for index, row in test_df.iterrows():
        # Объединим вектор запроса с вектором изображения
        combined_vector = np.hstack((query_vector, row['resnet50_vector']))
        combined_vector_tensor = torch.tensor(combined_vector).float().unsqueeze(0).to(device)

        # Прямой проход по модели
        nn_model.eval()
        with torch.no_grad():
            prediction = nn_model(combined_vector_tensor).cpu().numpy().flatten()[0]

        # Обновим лучшее изображение, если текущий балл выше.
        if prediction > max_score:
            max_score = prediction
            best_image = row['image']

    # Отображение изображения или дисклеймера
    if best_image is not None:
        img_path = os.path.join(test_images_dir, best_image)
        img = Image.open(img_path)
        plt.imshow(img)
        plt.show()
        return best_image, max_score
    else:
        print("No best image found or the request is incorrect.")
        return None, None

# Применим функцию к 10 запросам
random_queries = ["running boy in green shorts", 
                  "boxing",
                  "children",
                  "playing a game",
                  "girl with glass on the mountain",
                  "ice",
                  "sit on the beach",
                  "take pictures of sights",
                  "crowded fair",
                  "swim in the fountain"]
for query in random_queries:
    image, score = find_similar_image(query, test_df, nn_model, model, tokenizer, resnet_model, device)
    print(f"Query: {query}, Image: {image}, Score: {score}")

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Посмотри внимательно на функцию. Она НИКАК не использует полученный запрос (`query`), поэтому и просто  выдает, одну пару, для которой соответствие между изображением и описанием максимально.  </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Во первых, пожалуйста убери следы предсказаний (которых ты делаешь аж 500 штук при каждом запуске функции). Сделать это можно с помощью аргумента verbose метода predict.
    
Во вторых, не стоит 100/500 раз получать предсказание для каждой пары, лучше один раз запустить модель для 100 пар и получить оценку за один проход.    
  
</div>